### ISSUE TITLE
Piecewise non-constant quantile implementation

### ISSUE THREAD

**Issue by aimalz** [LSSTDESC/qp#8]
As promised in #5, I'm making an issue and corresponding branch for implementing a `qp.pdf_gen` subclass that's a variant of the piecewise constant interpolated `qp.quant_gen` to be consistent with the linearly interpolated behavior of [the previous version](https://github.com/aimalz/qp/blob/master/qp/pdf.py#L195) used for [the paper](https://arxiv.org/abs/1806.00014).

**Comment by aimalz** [comment]
Eureka! @johannct and I did some archaeology and figured out which pieces of the old qp belong in [https://github.com/LSSTDESC/qp/blob/master/qp/quant_pdf.py](`quant_pdf.py`). 

> The guts are here: https://github.com/aimalz/qp/blob/master/qp/pdf.py#L646
> Basically what I did before was manually create x values for 0% and 100% such that y(x(q=0))=0 and y(x(q=1))=0.
> Those define an extra interval on each end of the provided quantiles (x(q=0), x(q=qmin)) and (x(q=qmax), x(q=1)).
> I then solved for the y values defining triangles over those x intervals such that the areas of each are qmin and 1-qmax respectively, here: https://github.com/aimalz/qp/blob/master/qp/utils.py#L276
> That gave me y(x(q=qmin)) and y(x(q=qmax)). 
> Then I knew that the area between any x(q\_i) and x(q\_(i+1)) was equal to (q\_(i+1)) - q\_i so could solve for y(x(q\_(i+/-1))) given y(x(q\_i)), which is here: https://github.com/aimalz/qp/blob/master/qp/utils.py#L307

tl;dr it looks more or less like it can be copied wholesale from the old qp, aside from figuring out some vectorization stuff. 

(Also, I'd be lying if I didn't say that this has been a valuable lesson in documentation for users (docstrings explaining the types of all inputs and outputs) that I'm pretty good about emphasizing and documentation for developers (comments explaining what each bit of obtuse code is actually doing) that I rarely think about until I'm trying to remember what I did 5+ years ago.)

**Comment by aimalz** [comment]
From @johannct, don't forget to refactor these kinds of lines:
```
in_inds = ((xf >= self.quantiles[1][0]) & (xf <= self.quantiles[1][-1])).nonzero()[0]
lo_inds = ((xf < self.quantiles[1][0]) & (xf >= z[0])).nonzero()[0]
hi_inds = ((xf > self.quantiles[1][-1]) & (xf <= z[-1])).nonzero()[0]
```

**Comment by aimalz** [comment]
I really don't think the current solution that lacks enforcing continuity at the piecewise boundaries suffices. In @sschmidt23 's examples, it wouldn't satisfy the definition of quantiles either, in terms of area under the curve.

**Comment by drewoldag** [comment]
I've put in a fair amount of time trying to understand the history of this issue, and determine what it is that needs to be done here.  There are a few different levels of modifications that I can imagine being reasonable here, but at the same time, I'm not exactly sure which are appropriate and which are not. 

I've gone over the Slack conversations from October 5-6 2021, it seemed there was a fair amount of debate about the correct mathematical implementation of piecewise linear quantile parameterization. Due to the fact that those were historical conversations, and only some of the folks who had them were around to talk to (due to vacations) I've done my best to piece things together. 

<< End of preamble >> 

Some things I've noticed:

1. Within the `quant_gen._pdf` function, it seems that we can use `interpolate_multi_x_multi_y`.
2. Because we are using a forward (or backward) difference to calculate the numerical derivative, we need to offset the x values when passing them to `interpolate_multi_x_multi_y` by half the step size. ([Stack overflow link](https://stackoverflow.com/questions/17937979/calculating-the-derivative-of-cumulative-density-function-in-python))
Introducing the x offset, makes the PDF look a lot better.

Baseline:
<img width="621" alt="Screen Shot 2022-12-21 at 4 51 25 PM" src="https://user-images.githubusercontent.com/47493171/209031220-f3ae12a4-672f-440d-8435-47960669870d.png">
Including x offset:
<img width="621" alt="Screen Shot 2022-12-21 at 4 51 03 PM" src="https://user-images.githubusercontent.com/47493171/209031178-007ace53-fff8-49b1-9269-90bcb5d518d9.png">

3. Another potential option is to make use of np.gradient which takes the central difference, and eliminates the need for the x offset. It also reduces the amount of code to maintain. It handles the boundaries and non-uniform grids too. But I'll need assistance from the researches to know if it will play nicely with `pad_quantiles`, and whether it meets the criteria necessary (integrates to 1, continuous, etc.) 
4. The fundamental problem with the approach to quantile parameterization, from what I've seen, is the way that (quantile, location) pairs are calculated. For mundane examples, a uniform grid of quantiles will produce a nice representation of the original PDF. But for peaky, multimodal PDFs, this approach fails. There are sections of the CDF that are not well sampled, and this leads to the wings that are typically seen like in issue #65. I'll leave one potential solution in the next comment. 

At the end of the day though, it's not clear to me, without additional input from the researchers, what corrections need to be made to satisfy the mathematical requirements as well as the requirements of this ticket :)

**Comment by drewoldag** [comment]
Addressing point 4 in my last comment, one approach is to sample both linearly along quantiles, and then again along CDF. Then merge the two sets of (quantile, CDF) pairs, sort, and eliminate duplicates.  Forgive the screenshots, but what's shown below is an example with a 3 peak input PDF, the corresponding CDF with quantiles shown as blue dots, and the resulting PDF of the quantile parameterization. 

The second row shows that doubling the number of quantiles doesn't meaningfully increase the quality of the qp PDF. 

In the third row, I take a second sample of CDF values (linearally interpolated in this case) and combining those with the previous (quantile, CDF) pairs, produces a much better qp PDF. 

<img width="1190" alt="Screen Shot 2022-12-21 at 5 00 48 PM" src="https://user-images.githubusercontent.com/47493171/209032158-9cebe635-f283-4007-9f4e-1887b0fa6df0.png">

Here's the pseudo code:
```
# Create the standard (quantile, CDF) pairs - CDFs, quantiles are the blue dots on the CDF plot above
quantiles = np.linspace(0.01, 0.99, 51)
CDFs = distribution.ppf(quantiles)

# Define a second range of CDFs - other_CDFs, other_quantiles are the orange dots on the CDF plot above
other_CDFs = np.linspace(min(CDFs), max(CDFs), 51)
interp_function = interp1d(CDFs, quantiles, kind='linear')
other_quantiles = interp_function(other_CDFs)

# Merge the two sets of (quantile, CDFs) pairs together
all_CDFs = np.concatenate((CDFs, other_CDFs))
all_quantiles = np.concatenate((quantiles, other_quantiles))

# Get indexes for sorting
inds = np.argsort(all_pcts)

# Sort the two arrays so that all_CDFs is always increasing
combined_pcts = np.take_along_axis(all_pcts, inds, axis=0)
combined_percentiles = np.take_along_axis(all_percentiles, inds, axis=0)

# Remove any duplicate values (duplicates will result in “divide by 0” errors)
output_pcts, inds = np.unique(combined_pcts, return_index=True)
output_percentiles = np.take_along_axis(combined_percentiles, inds, axis=0)
```

**Comment by drewoldag** [comment]
Final comment on this, comparing the piecewise constant implementation against the piecewise linear with second sampling pass.

<img width="1060" alt="Screen Shot 2022-12-21 at 5 04 33 PM" src="https://user-images.githubusercontent.com/47493171/209032521-7c16e132-acac-40da-bff4-c415e617f43c.png">

**Comment by sschmidt23** [comment]
(I'm already on break but will comment briefly, and I obviously don't expect a response if anyone else is also already on vacation): I like this solution of basically ensuring that there aren't large stretches of redshift with no quantile, a nice fix to the problem that we were seeing with the initial implementation.  Did you do any checks on how much more computation is needed for this new method vs the piecewise quantile method?  The piecewise and x/y sampled look qualitatively similar in performance in your final plot above, I'm wondering if we think that any extra computational expense is worth the difference?  I guess now that both are implemented that point is moot, the user can decide to use which of the two methods best meets their needs, I am just curious what the relative computational load is for the two methods.

**Comment by drewoldag** [comment]
I'm really glad that you asked the question, I have an answer that is based on what I've seen, but I haven't gone too deep into the details to quantify it yet, because I wasn't sure that the approach was reasonable. 

For the moment, if we ignore the calculation of the (quantile, CDF) pairs, and just look at the quant_gen vs. quant_piecewise_gen code, they are fairly similar. quant_piecewise_gen has a method, `_valatloc` that is calculating the derivative, but isn't calling it that. And `quant_gen` has a method `_compute_derivs` doing the same thing. 

In fact, it's a duplicated line of code here: https://github.com/LSSTDESC/qp/blob/main/src/qp/quant_pdf.py#L123 and here: https://github.com/LSSTDESC/qp/blob/main/src/qp/quant_pdf.py#L261

The non-constant implementation was calculating an additional difference, but I don't think that's required at this point. Therefore, from that perspective, the calculations are basically identical. 

Looking at the interpolation that happens when calling `*._pdf`:
- In `quant_gen` we make a call ultimately to `scipy.interpolation.interp1d`. Given the parameters that we're passing in, there is a merge sort that interp1d call, that we could avoid since our inputs are already sorted. And it looks like once it's sorted in `interp1d` it makes a call to `numpy.interp` which is a pretty tight set of calculations. (In fact, we could probably replace `scipy.interpolation.interp1d` with `numpy.interp`.)
- In `quant_peicewise_gen` there's a call to `evaluate_hist_multi_x_multi_y_product` which performs a few lines of numpy-based calculations, so presumably those are fast. 

All that being said, by eye, it looks like the performance should be comparable across `quant_gen` and `quant_piecewise_gen`. I can't immediately tell what the perf. difference will be when comparing the different methods of gathering (quantile, CDF) pairs. 

Investigating that would be my next step - again, I just didn't want to sink a lot of time into it, if the idea was fundamentally flawed.

**Comment by drewoldag** [comment]
For a single distribution, this approach of sampling once in quantiles and again in CDF works pretty nicely. However, from a storage perspective (which I believe was the original intent of the technique) it starts to suffer when we think about ensembles. 

In the original implementation, a single array of Q quantiles (Q = a few dozen) could be defined for N distributions (N = many 10's, 1000's or more) stored in an ensemble each with an array of C locations (dimension of C = dimension of Q). The same Q quantiles we applicable for every distribution, and only the C's were different.

If we take this approach of sampling CDFs, it means that now we must also store some number of unique Q's for each distribution along with all the C values. 

### Some quick math
Considering the baseline approach with 100 quantiles and 1000 distributions in an ensemble. The total number of values to be stored would be: 
100 quantiles + 1000 dists * 100 location values = **100100 values**

Considering the proposed approach 50 reusable quantiles and 50 quantiles specific to 1000 distributions. The total number of values to be stored would be: 
50 quantiles + 1000 dists * (100 locations values + 50 dist-specific quantiles) = **150050 values**

### Initial conclusions
Cons:
Work will be required in the quant_gen class to manipulate the new array shapes. There will be a trade off of complexity of implementation and efficiency of storage.
Extra memory and storage space required no matter what. Roughly 1.5-2x more than the baseline.

Neutral:
No computational penalty for generating or working with the piecewise linear classes.
Not clear if there is a significant improvement in the fidelity of the PDFs that are produced over the piecewise constant approach.

Pros:
PDF output looks a lot nicer for wider, single peak PDFs than the piecewise constant implementation.
